src/core/mac/sub_mac_wed.cpp

```diff
@@ -0,0 +1,87 @@
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the Wake-up End Device of the subset of IEEE 802.15.4 MAC primitives.
+ */
+
+#include "sub_mac.hpp"
+
+#if OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+
+#include "instance/instance.hpp"
+
+namespace ot {
+namespace Mac {
+
+RegisterLogModule("SubMac");
+
+void SubMac::WedInit(void)
+{
+    mWedListenInterval = 0;
+    mWedTimer.Stop();
+}
+
+void SubMac::UpdateWakeupListening(uint32_t aInterval, uint16_t aDuration, uint8_t aChannel)
+{
+    OT_ASSERT(RadioSupportsReceiveTiming());
+
+    VerifyOrExit(aInterval != mWedListenInterval || aDuration != mWedListenDuration || aChannel != mWakeupChannel);
+
+    mWedListenInterval = aInterval;
+    mWedListenDuration = aDuration;
+    mWakeupChannel     = aChannel;
+    mWedTimer.Stop();
+
+    if (mWedListenInterval > 0)
+    {
+        mWedSampleTime = TimeMicro(otPlatRadioGetNow(&GetInstance())) + kCslReceiveTimeAhead;
+        HandleWedTimer();
+    }
+
+exit:
+    return;
+}
+
+void SubMac::HandleWedTimer(Timer &aTimer) { aTimer.Get<SubMac>().HandleWedTimer(); }
+
+void SubMac::HandleWedTimer(void)
+{
+    if (RadioSupportsReceiveTiming() && (mState != kStateDisabled))
+    {
+        IgnoreError(Get<Radio>().ReceiveAt(mWakeupChannel, mWedSampleTime.GetValue(), mWedListenDuration));
+    }
+    mWedSampleTime += mWedListenInterval;
+    mWedTimer.FireAt(mWedSampleTime - kCslReceiveTimeAhead);
```

Comment:
Regarding `mWedSampleTime`, we need to be careful about the time basis used, since we have two different notions of time:

- `otPlatRadioGetNow()`, which provides the radio time (a `uint64_t` value).
- `otPlatAlarm{Micro/Milli}GetNow()`, which map to `Timer{Milli/Micro}::GetNow()` and use `uint32_t` values.

The `mWedTimer` which is a `TimerMicro` works with the latter, but the `Radio::ReceiveAt()` API assumes its `aStart` input is based on `otPlatRadioGetNow()`.

Therefore, either `mWedTimer.FireAt(mWedSampleTime)` or `ReceiveAt(mWedSampleTime.GetValue())` might be incorrect. 

On some platforms, these two time sources might be the same, but we need to consider that they may differ and adjust for any difference between them.

----

Here is my suggestion:

- We can base `mWedSampleTime`, which is a `TimeMicro` object, on `TimeMicro::GetNow()`.
- This allows us to use it when starting the `mWedTimer`.
- When calling `ReceiveAt()`, convert the time to the radio time, something like:
```c++
     static_cast<uint32_t>(otPlatRadioGetNow()) + (mWedSampleTime - TimerMicro::GetNow())`
```

A similar situation may apply to the CSL code, so adding some helper functions to convert between the two time sources could be useful.

Reply:
That's right, however using radio API for every interval is costly for some platforms. I'll address the issue by maintaining both local and radio times separately.